### DISCUSSION TITLE
Any way to add custom banners on versioned docs?

### DISCUSSION THREAD

**Discussion by andrewnicols** [facebook/docusaurus#7112] [Q&A]
I've [raised a feature request](https://docusaurus.io/feature-requests/p/feature-request-custom-banner-types-for-versioned-docs) on canny for this functionality, but am also wondering whether there's a way to implement this in the mean time through some combination of a custom plugin, swizzle, or ..?

In our project we use the `master` branch for our next version. We also support one-two stable versions for 18 months, at which point they move into a security status, and some of them enter an LTS state with extended security support for 3 years.

Ideally we would like to display that information in the docs in the version banner, with text to indicate something like "This documentation relates to a security relase of Moodle. For the latest docs see [master link]" or similar.

I see that the current list of banners is validated in `node_modules/@docusaurus/plugin-content-docs/src/options.ts`, so even if I swizzle the `DocsVersionBanner` (unsafe) to possibly add additional banners, their use will be prevented by the configuration validation.

Is there a way around this, short of extending `plugin-content-docs` and (somehow??) replacing the options.ts. I'm new to both React and Typescript, so this is a learning curve for me and any pointers would be appreciated.

Thanks

**Comment by andrewnicols** [comment]
As one possible API within the configuration file:
```
                    versionBanners: {
                        security: {
                            id:  'theme.docs.versions.securityVersionLabel',
                            translation: 'This is documentation for {siteTitle} {versionLabel}, which is in a security supported state.',
                        },
                        stable: {
                            id:  'theme.docs.versions.stableVersionLabel',
                            translation: 'This is documentation for {siteTitle} {versionLabel}, which is a stable supported state.',
                        },
                        lts: {
                            id:  'theme.docs.versions.ltsVersionLabel',
                            translation: 'This is documentation for {siteTitle} {versionLabel}, which is in a long-term support state.',
                        },
                    },
                    version: {
                        current: {
                            label: 'master',
                            banner: 'none',
                        },
                        '4.0': {
                            banner: 'stable',
                        },
                        '3.11': {
                            banner: 'stable',
                        },
                        '3.10': {
                            banner: 'security',
                        },
                        '3.9': {
                            banner: 'lts',
                        },
                    },
```

**Comment by slorber** [comment]
Hey

This may not be ideal but you could swizzle the component, use `useDocsVersion()` and hardcode the different cases in the component directly depending on the version?

I guess we could also want to implement these 2 things later:
- Ability to assign custom data to a version through config (not necessarily any version banner type though, because we want to prevent config typos)
- Ability to "register" your own banner types (we'll likely want to do that too for custom navbar, footer, sidebar items etc...)

**Comment by withshubh** [comment]
Hey @andrewnicols - Thanks for creating this discussion. Is there any update on it? Any issue that I can follow?
I was trying to look for a way to add a custom banner text on the versioned doc, did you happen to figure out any workaround?

**Comment by debbly** [comment]
Chiming in - also looking for a solution I can check out around custom banners!

**Reply by slorber** [reply]
That would help design the feature if you detailed the need for custom banners.

I was thinking that a lower-level primitive we could provide, less convenient but more powerful, could be to let you assign custom metadata to a version, so that it becomes accessible in docs swizzled components? 

This way you could implement custom banners, or whatever else you had in mind that is not necessarily a banner.

---

This could be done by extending the versions config in docs plugin option:
https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs

```js
const docsConfig = {
  /* ... */
  versions: {
    current: {
      label: "Android SDK v2.0.0 (WIP)",
      path: "android-2.0.0",
      banner: "none",
      customProps: {
        myCustomBanner: "whatever you want"
      }
    },
    "1.0.0": {
      label: "Android SDK v1.0.0",
      path: "android-1.0.0",
      banner: "unmaintained",
      customProps: {
        myCustomBanner: "whatever you want"
      }
    },
  },
};
```

We already support this for sidebar items and it unlocks quite a few use-cases.

**Comment by andrewnicols** [comment]
I went with @slorber 's suggested approach in the end. This may change when I look to migrate to v3.

- [Remark plugin to include the `MoodlePageBanner`](https://github.com/moodle/devdocs/blob/main/src/remark/moodleBanner.js)
- [Docusaurus config to include the remark plugin](https://github.com/moodle/devdocs/blob/main/docusaurus.config.js#L34). 
- [The `MoodlePageBanner` which is actually a container adding other types of banner](https://github.com/moodle/devdocs/blob/main/src/theme/MoodlePageBanner/index.js):
  - [Unsupported Version banner](https://github.com/moodle/devdocs/blob/main/src/theme/UnsupportedVersionBanner/index.js)
  - [Draft Documentation banner](https://github.com/moodle/devdocs/blob/main/src/theme/DraftPageBanner/index.js) - setting {{draft: true}} in frontpage matter prevents the page from being generated at al. We want the docs generated but with a CTA to help complete/update/fix them for specific pages

We will probably keep this approach for now because:
- it's based on the [release data for our project](https://github.com/moodle/devdocs/blob/main/data/versions.json) so we have a single source of truth. Each release has a date, and we use that to determine if the release is unsupported, in security-only support, supported, or a future release
- we found the need for a draft banner to add a CTA for help

We can probably move the version banner to config fairly easily and keep the data-driven approach, but the draft banner is also incredibly helpful. I might look into raising a PR for that if anyone else is interested in adding a similar thing via frontpage matter.
